File: leetcode_python/Array/count-number-of-nice-subarrays.py
"""

1248. Count Number of Nice Subarrays
Medium

Given an array of integers nums and an integer k. A continuous subarray is called nice if there are k odd numbers on it.

Return the number of nice sub-arrays.

 

Example 1:

Input: nums = [1,1,2,1,1], k = 3
Output: 2
Explanation: The only sub-arrays with 3 odd numbers are [1,1,2,1] and [1,2,1,1].
Example 2:

Input: nums = [2,4,6], k = 1
Output: 0
Explanation: There is no odd numbers in the array.
Example 3:

Input: nums = [2,2,2,1,2,2,1,2,2,2], k = 2
Output: 16
 

Constraints:

1 <= nums.length <= 50000
1 <= nums[i] <= 10^5
1 <= k <= nums.length

"""

# V0
# IDEA : cumsum + dict (Prefix sum)
class Solution:
    def numberOfSubarrays(self, nums, k):
        d = collections.defaultdict(int)
        d[0] = 1
        cur_sum = 0
        res = 0
        for i in nums:
            if i % 2 == 1:
                cur_sum += 1
            if cur_sum - k in d:
                res += d[cur_sum - k]
            d[cur_sum] += 1
        return res

# V1
# IDEA : cumsum + dict (Prefix sum)
# https://leetcode.com/problems/count-number-of-nice-subarrays/discuss/868119/python-solution
class Solution:
    def numberOfSubarrays(self, nums, k):
        d = collections.defaultdict(int)
        d[0] = 1
        cur_sum = 0
        res = 0
        for i in nums:
            if i % 2 == 1:
                cur_sum += 1
            if cur_sum - k in d:
                res += d[cur_sum - k]
            d[cur_sum] += 1
        return res

# V1''
# IDEA : exact k
# https://leetcode.com/problems/count-number-of-nice-subarrays/discuss/419668/Python-3-ways-summarize
# IDEA :
# Convert "exact k" to "at most k". O(n) time O(1) space
# Lee215's way. We are looking for result of "exact k", we can calculate "at most k" first, and final result will be: "exact k" == "atmost k" - "atmost k-1".
class Solution:
    def numberOfSubarrays(self, nums: List[int], k: int) -> int:
            l=len(nums)
            def atmost(k):
                res=0
                cnt=0
                j=0
                for i in range(l):
                    if nums[i]%2==1:
                        cnt+=1
                    while cnt>k:
                        if nums[j]%2==1:
                            cnt-=1
                        j+=1
                    res+=i-j+1
                return res
            return atmost(k)-atmost(k-1)

# V1'''
# IDEA : deque
# https://leetcode.com/problems/count-number-of-nice-subarrays/discuss/419668/Python-3-ways-summarize
# IDEA :
# Record all odd indexes. O(n) time O(n) space
# Record indexes of all odd numbers in a deque. If len(que)==k+1, that means we have k+1 odd numbers now, all subarray starting between que[0] and que[1], ending at current index i, are valid sub arrays.
class Solution:
    def numberOfSubarrays(self, nums: List[int], k: int) -> int:
            res=0
            que=collections.deque([-1])
            for i,n in enumerate(nums):
                if n%2==1:
                    que.append(i)
                if len(que)==k+2:
                    que.popleft()
                if len(que)==k+1:
                    res+=que[1]-que[0]
            return res

# V1'''''
# IDEA : prefix sum
# https://leetcode.com/problems/count-number-of-nice-subarrays/discuss/419668/Python-3-ways-summarize
# IDEA
# Prefix sum. O(n) time O(n) space
# Odd number being 1, even number being 0, calculate prefix sum of the whole nums. "visited" is a dictionary, key being prefix, value being the repeating count of that prefix.
# Prefix sum is an increasing sequence, so visited[p] will always represent the prefix happened before current index.
class Solution:
    def numberOfSubarrays(self, nums: List[int], k: int) -> int:
            res=cursum=0
            visited=collections.defaultdict(int)
            for v in nums:
                visited[cursum]+=1
                cursum+=v%2
                if cursum-k in visited:
                    res+=visited[cursum-k]
            return res

# V1'''''''''
# https://leetcode.com/problems/count-number-of-nice-subarrays/discuss/421157/Simple-Python-solution
class Solution:
    def numberOfSubarrays(self, nums, k):
        memo = {0:-1}
        ans = cur = 0
        for i in range(len(nums)):
            if nums[i] % 2:
                cur += 1
                memo[cur] = i
            if cur >= k:
                ans += memo[cur + 1 - k] - memo[cur - k]
        return ans

# V1'''''''''
# IDEA : LC 828 + windows
# https://leetcode.com/problems/count-number-of-nice-subarrays/discuss/421003/Python-straightforward-solution
# https://leetcode-cn.com/problems/count-number-of-nice-subarrays/
# IDEA :  Straightforward Solution: similar to the idea in the solution for 828. Unique Letter String
class Solution:
    def numberOfSubarrays(self, nums, k):
            lst = [-1]
            for i in range(len(nums)):
                if nums[i] % 2:
                    lst.append(i)
            lst.append(len(nums))
            res = 0
            for i in range(1, len(lst) - k):
                # plus the number of windows containing [lst[i], lst[i+k-1]]
                res += (lst[i] - lst[i-1]) * (lst[i+k] - lst[i+k-1])
            return res

# V1''''''''''
# https://leetcode.com/problems/count-number-of-nice-subarrays/discuss/420390/Simple-Python-Solution
class Solution:
    def numberOfSubarrays(self, nums, k) :
        odd = [i for i in range(len(nums)) if nums[i] % 2 == 1]
        if len(odd) < k: return 0
        res = 0
        odd = [-1] + odd + [len(nums)]
        for i in range(1, len(odd) - k):
            res += (odd[i]- odd[i-1]) * (odd[i + k] - odd[i + k - 1])
        return res

# V1'''''''''''
# IDEA : AT MOST
# https://leetcode.com/problems/count-number-of-nice-subarrays/discuss/419378/JavaC%2B%2BPython-Sliding-Window-O(1)-Space
class Solution:
    def numberOfSubarrays(self, A, k):
        def atMost(k):
            res = i = 0
            for j in range(len(A)):
                k -= A[j] % 2
                while k < 0:
                    k += A[i] % 2
                    i += 1
                res += j - i + 1
            return res

        return atMost(k) - atMost(k - 1)

# V1''''''''''''
# IDEA : ONE PASS
# https://leetcode.com/problems/count-number-of-nice-subarrays/discuss/419378/JavaC%2B%2BPython-Sliding-Window-O(1)-Space
class Solution:
    def numberOfSubarrays(self, A, k):
        i = count = res = 0
        for j in range(len(A)):
            if A[j] & 1:
                k -= 1
                count = 0
            while k == 0:
                k += A[i] & 1
                i += 1
                count += 1
            res += count
        return res

# V1''''''''''''''
# https://littlebees.github.io/2021/08/leetcode-1248/
class Solution:
    def numberOfSubarrays(self, nums: List[int], k: int) -> int:
        def atmost(limit):
            i = cnt = ret = 0
            for j,n in enumerate(nums):
                if n % 2 == 1:
                    cnt += 1
                while i <= j and cnt > limit:
                    if nums[i] % 2 == 1:
                        cnt -= 1
                    i += 1
                ret += j-i+1
            return ret
        return atmost(k)-atmost(k-1)

# V1'''''''''''''''
# https://littlebees.github.io/2021/08/leetcode-1248/
class Solution:
    def numberOfSubarrays(self, A, k):
        i = count = res = 0
        for j in range(len(A)):
            if A[j] & 1:
                k -= 1
                count = 0 # get len till meet the condition
            while k == 0:
                k += A[i] & 1
                i += 1
                count += 1
            res += count
        return res

# V2

File: leetcode_java/src/main/java/LeetCodeJava/Array/CountNumberOfNiceSubarrays.java
package LeetCodeJava.Array;

// https://leetcode.com/problems/count-number-of-nice-subarrays/description/

import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;
import java.util.Queue;

/**
 * 1248. Count Number of Nice Subarrays
 * Solved
 * Medium
 * Topics
 * premium lock icon
 * Companies
 * Hint
 * Given an array of integers nums and an integer k. A continuous subarray is called nice if there are k odd numbers on it.
 *
 * Return the number of nice sub-arrays.
 *
 *
 *
 * Example 1:
 *
 * Input: nums = [1,1,2,1,1], k = 3
 * Output: 2
 * Explanation: The only sub-arrays with 3 odd numbers are [1,1,2,1] and [1,2,1,1].
 * Example 2:
 *
 * Input: nums = [2,4,6], k = 1
 * Output: 0
 * Explanation: There are no odd numbers in the array.
 * Example 3:
 *
 * Input: nums = [2,2,2,1,2,2,1,2,2,2], k = 2
 * Output: 16
 *
 *
 * Constraints:
 *
 * 1 <= nums.length <= 50000
 * 1 <= nums[i] <= 10^5
 * 1 <= k <= nums.length
 *
 */
public class CountNumberOfNiceSubarrays {

    // V0
//    public int numberOfSubarrays(int[] nums, int k) {
//
//    }



    // V0-1
    // IDEA: PREFIX SUM + HASHMAP (GPT)
    /**
     *
     * ## 📊 Walkthrough Example
     *
     * ```
     * nums = [1,1,2,1], k = 2
     * ```
     *
     * | step | num | oddCount | map               | res |
     * | ---- | --- | -------- | ----------------- | --- |
     * | init |     | 0        | {0:1}             | 0   |
     * | 1    | 1   | 1        | {0:1,1:1}         | 0   |
     * | 2    | 1   | 2        | {0:1,1:1,2:1}     | 1   |
     * | 3    | 2   | 2        | {0:1,1:1,2:2}     | 2   |
     * | 4    | 1   | 3        | {0:1,1:1,2:2,3:1} | 4   |
     *
     * ---
     *
     * ## 🔑 Intuition in One Line
     *
     * HashMap helps you answer:
     *
     * > “How many valid starting points exist for the current ending point?”
     *
     * ---
     *
     * ## ⚖️ Sliding Window vs HashMap
     *
     * | Approach         | When to use       |
     * | ---------------- | ----------------- |
     * | Sliding Window   | at most / longest |
     * | Prefix + HashMap | exact count       |
     *
     */
    public int numberOfSubarrays_0_1(int[] nums, int k) {

        /**  NOTE !!!
         *
         *  map :  {oddCount : frequency }
         *
         *  -> How many times a certain number
         *     of odd numbers has appeared so far
         */
        Map<Integer, Integer> map = new HashMap<>();
        // NOTE !!! init as {0, 1}
        map.put(0, 1);

        int res = 0, oddCount = 0;

        for (int num : nums) {
            if (num % 2 == 1) {
                oddCount++;
            }

            /** NOTE !!!
             *
             *  “How many previous positions had exactly oddCount - k odds?”
             *
             *
             *  ->
             *
             *  - If current = oddCount
             *     -> Previous = oddCount - k
             *     -> Then subarray between them has exactly k odds
             *
             */
            res += map.getOrDefault(oddCount - k, 0);
            map.put(oddCount, map.getOrDefault(oddCount, 0) + 1);
        }

        return res;
    }


    // V0-1
    // IDEA: Sliding Window (GPT)
    /** NOTE !!!
     *
     *   Core idea:
     *
     *   (instead of Instead of counting directly, use below)
     *
     *   # of subarrays with exactly k odds =
     *        atMost(k) - atMost(k - 1)
     *
     */
    public int numberOfSubarrays_0_2(int[] nums, int k) {
        return atMost_0_1(nums, k) - atMost_0_1(nums, k - 1);
    }

    private int atMost_0_1(int[] nums, int k) {
        int l = 0, res = 0, oddCount = 0;

        for (int r = 0; r < nums.length; r++) {
            if (nums[r] % 2 == 1) {
                oddCount++;
            }

            while (oddCount > k) {
                if (nums[l] % 2 == 1) {
                    oddCount--;
                }
                l++;
            }

            // count subarrays ending at r
            res += (r - l + 1);
        }

        return res;
    }


    // V0-3
    // IDEA: Sliding Window (GEMINI)
    /** NOTE !!!
     *
     *   Core idea:
     *
     *   (instead of Instead of counting directly, use below)
     *
     *   # of subarrays with exactly k odds =
     *        atMost(k) - atMost(k - 1)
     *
     */
    public int numberOfSubarrays(int[] nums, int k) {
        // Exactly k = (At most k) - (At most k-1)
        return atMost_0_3(nums, k) - atMost_0_3(nums, k - 1);
    }

    private int atMost_0_3(int[] nums, int k) {
        if (k < 0)
            return 0;
        int l = 0, count = 0, res = 0;

        for (int r = 0; r < nums.length; r++) {
            // If the number is odd, it contributes to our 'k' count
            if (nums[r] % 2 == 1) {
                k--;
            }

            // Shrink window if we exceed k odd numbers
            while (k < 0) {
                if (nums[l] % 2 == 1) {
                    k++;
                }
                l++;
            }

            // The number of subarrays ending at 'r' with AT MOST k odds
            // is exactly the length of the current window.
            res += (r - l + 1);
        }
        return res;
    }



    // V1-1
    // IDEA: HASHING
    // https://leetcode.com/problems/count-number-of-nice-subarrays/editorial/
    public int numberOfSubarrays_1_1(int[] nums, int k) {
        int currSum = 0, subarrays = 0;
        Map<Integer, Integer> prefixSum = new HashMap<>();
        prefixSum.put(currSum, 1);

        for (int i = 0; i < nums.length; i++) {
            currSum += nums[i] % 2;
            // Find subarrays with sum k ending at i
            if (prefixSum.containsKey(currSum - k)) {
                subarrays = subarrays + prefixSum.get(currSum - k);
            }
            // Increment the current prefix sum in hashmap
            prefixSum.put(currSum, prefixSum.getOrDefault(currSum, 0) + 1);
        }

        return subarrays;
    }


    // V1-2
    // IDEA: Sliding Window using Queue
    // https://leetcode.com/problems/count-number-of-nice-subarrays/editorial/
    public int numberOfSubarrays_1_2(int[] nums, int k) {
        Queue<Integer> oddIndices = new LinkedList<>();
        int subarrays = 0;
        int lastPopped = -1;
        int initialGap = 0;

        for (int i = 0; i < nums.length; i++) {
            // If element is odd, append its index to the list.
            if (nums[i] % 2 == 1) {
                oddIndices.offer(i);
            }
            // If the number of odd numbers exceeds k, remove the first odd index.
            if (oddIndices.size() > k) {
                lastPopped = oddIndices.poll();
            }
            // If there are exactly k odd numbers, add the number of even numbers
            // in the beginning of the subarray to the result.
            if (oddIndices.size() == k) {
                initialGap = oddIndices.element() - lastPopped;
                subarrays += initialGap;
            }
        }

        return subarrays;
    }


    // V1-3
    // IDEA: Sliding Window (Space Optimisation of queue-based approach)
    // https://leetcode.com/problems/count-number-of-nice-subarrays/editorial/
    public int numberOfSubarrays_1_3(int[] nums, int k) {
        int subarrays = 0, initialGap = 0, qsize = 0, start = 0;
        for (int end = 0; end < nums.length; end++) {
            // If current element is odd, increment qsize by 1.
            if (nums[end] % 2 == 1) {
                qsize++;
            }
            // If qsize is k, calculate initialGap and add it in the answer.
            if (qsize == k) {
                initialGap = 0;
                // Calculate the number of even elements in the beginning of
                // subarray.
                while (qsize == k) {
                    qsize -= nums[start] % 2;
                    initialGap++;
                    start++;
                }
            }
            subarrays += initialGap;
        }
        return subarrays;
    }


    // V1-4
    // IDEA:  Sliding Window (subarray sum at most k)
    // https://leetcode.com/problems/count-number-of-nice-subarrays/editorial/
    public int numberOfSubarrays_1_4(int[] nums, int k) {
        return atMost(nums, k) - atMost(nums, k - 1);
    }

    private int atMost(int[] nums, int k) {
        int windowSize = 0, subarrays = 0, start = 0;

        for (int end = 0; end < nums.length; end++) {
            windowSize += nums[end] % 2;
            // Find the first index start where the window has exactly k odd elements.
            while (windowSize > k) {
                windowSize -= nums[start] % 2;
                start++;
            }
            // Increment number of subarrays with end - start + 1.
            subarrays += end - start + 1;
        }

        return subarrays;
    }



    // V2




}
